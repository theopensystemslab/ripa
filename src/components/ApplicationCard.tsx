import { makeStyles } from "@material-ui/core";
import Box from "@material-ui/core/Box";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import IconButton from "@material-ui/core/IconButton";
import ListItem from "@material-ui/core/ListItem";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import ListItemText from "@material-ui/core/ListItemText";
import Menu from "@material-ui/core/Menu";
import MenuItem from "@material-ui/core/MenuItem";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import { formatDistance } from "date-fns";
import * as React from "react";
import { CornerRightUp, Folder, Trash } from "react-feather";

const useStyles = makeStyles(theme => ({
  applications: {
    paddingTop: theme.spacing(2),
    marginBottom: theme.spacing(2)
  },
  application: {
    width: "100%",
    [theme.breakpoints.up("sm")]: {
      width: 264
    },
    height: "100%",
    display: "flex",
    flexDirection: "column"
  },
  link: {
    textDecoration: "none"
  },
  list: {
    color: "currentColor"
  },
  divider: {
    backgroundColor: theme.palette.primary.contrastText
  },
  content: {
    flexGrow: 1
  },
  thumbnail: {
    height: 0,
    paddingTop: "56.25%",
    cursor: "pointer"
  },
  menu: {
    padding: 0
  },
  menuItem: {
    minWidth: "10rem",
    display: "flex",
    justifyContent: "space-between",
    backgroundColor: theme.palette.grey[800],
    color: "#fff",
    "&:hover": {
      backgroundColor: theme.palette.grey[900]
    }
  },
  listedApplication: {
    color: theme.palette.grey[900],
    backgroundColor: "rgba(255, 255, 255, 0.6)",
    "&:not(:last-child)": {
      marginBottom: 2
    }
  },
  panelButton: {
    display: "flex",
    width: "100%",
    fontSize: 17,
    justifyContent: "flex-start",
    padding: theme.spacing(2),
    fontWeight: 700,
    "&:hover": {
      backgroundColor: "rgba(0,0,0,0.035)"
    }
  },
  panelIcon: {
    minWidth: 0,
    marginRight: theme.spacing(2),
    color: "currentColor"
  },
  collapseIcon: {
    position: "absolute",
    right: theme.spacing(2),
    top: "50%",
    transition: "transform 0.3s ease-out",
    transform: "translateY(-50%)"
  },
  panelButtonActive: {
    backgroundColor: "rgba(0,0,0,0.035)"
  },
  collapseIconActive: {
    transform: "translateY(-50%) rotate(180deg)"
  }
}));

interface IApplication {
  thumbnail: string;
  updatedAt: number;
  description: string;
  status: string;
  listView?: boolean;
}

const ApplicationCard = (
  { thumbnail, description, updatedAt, status, listView }: IApplication,
  { ...props }
) => {
  const [anchorEl, setAnchorEl] = React.useState(null);

  const handleClick = event => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const classes = useStyles();
  if (!listView) {
    return (
      <Card
        data-testid="standardCard"
        className={classes.application}
        {...props}
      >
        <CardMedia className={classes.thumbnail} image={thumbnail} />
        <CardContent className={classes.content}>
          <Box fontSize="subtitle1.fontSize">
            <strong>{description}</strong>
          </Box>
          <Box fontSize="subtitle1.fontSize" color="grey.400" mb={3}>
            Last edited {formatDistance(updatedAt, new Date())} ago
          </Box>
        </CardContent>
        <CardActions>
          <Box
            pl={1}
            flexGrow={1}
            fontSize="subtitle1.fontSize"
            color={status === "Submitted" ? "#000" : "grey.500"}
          >
            {status}
          </Box>
          <IconButton data-testid="moreButtonStandard" onClick={handleClick}>
            <MoreVertIcon />
          </IconButton>
          <Menu
            id="application-menu"
            anchorEl={anchorEl}
            keepMounted
            open={Boolean(anchorEl)}
            onClose={handleClose}
            MenuListProps={{
              classes: {
                root: classes.menu
              }
            }}
            elevation={0}
            getContentAnchorEl={null}
            anchorOrigin={{
              vertical: "center",
              horizontal: "center"
            }}
            transformOrigin={{
              horizontal: "right",
              vertical: "top"
            }}
          >
            <MenuItem
              className={classes.menuItem}
              component="a"
              onClick={handleClose}
            >
              <ListItemText primary="Archive" />
              <Folder size={20} />
            </MenuItem>
            <MenuItem
              className={classes.menuItem}
              component="a"
              onClick={handleClose}
            >
              <ListItemText primary="Delete" />
              <Trash size={20} />
            </MenuItem>
          </Menu>
        </CardActions>
      </Card>
    );
  }
  return (
    <ListItem data-testid="listViewCard" className={classes.listedApplication}>
      <ListItemText
        primary={description}
        secondary={`Last edited ${formatDistance(updatedAt, new Date())} ago`}
      />
      <ListItemSecondaryAction>
        <IconButton
          data-testid="moreButtonList"
          onClick={handleClick}
          color="default"
        >
          <MoreVertIcon />
        </IconButton>
        <Menu
          id="application-menu"
          anchorEl={anchorEl}
          keepMounted
          open={Boolean(anchorEl)}
          onClose={handleClose}
          MenuListProps={{
            classes: {
              root: classes.menu
            }
          }}
          elevation={0}
          getContentAnchorEl={null}
          anchorOrigin={{
            vertical: "center",
            horizontal: "center"
          }}
          transformOrigin={{
            horizontal: "right",
            vertical: "top"
          }}
        >
          <MenuItem
            className={classes.menuItem}
            component="a"
            onClick={handleClose}
          >
            <ListItemText primary="Restore" />
            <CornerRightUp size={20} />
          </MenuItem>
          <MenuItem
            className={classes.menuItem}
            component="a"
            onClick={handleClose}
          >
            <ListItemText primary="Delete" />
            <Trash size={20} />
          </MenuItem>
        </Menu>
      </ListItemSecondaryAction>
    </ListItem>
  );
};

export default ApplicationCard;
