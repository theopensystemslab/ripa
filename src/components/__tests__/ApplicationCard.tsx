import { cleanup, fireEvent, render, wait } from "@testing-library/react";
import * as React from "react";

import ApplicationCard from "../ApplicationCard";

afterEach(cleanup);

describe("Application Card Component", () => {
  let thumbnail: string;
  let updatedAt: number;
  let description: string;
  let status: string;
  let listView: boolean;
  beforeAll(() => {
    thumbnail = "pic.png";
    description = "description";
    updatedAt = 7;
    status = "Submitted";
  });
  describe("of type list view", () => {
    it("should render snapshot", () => {
      listView = true;
      const { asFragment } = render(
        <ApplicationCard
          thumbnail={thumbnail}
          description={description}
          updatedAt={updatedAt}
          status={status}
          listView={listView}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("should open a menu when clicked", async () => {
      listView = true;
      const { getByTestId, findByText } = render(
        <ApplicationCard
          thumbnail={thumbnail}
          description={description}
          updatedAt={updatedAt}
          status={status}
          listView={listView}
        />
      );
      fireEvent.click(getByTestId("moreButtonList"));
      expect(await findByText(/restore/i)).toBeInTheDocument();
      expect(await findByText(/delete/i)).toBeInTheDocument();
    });
  });
  describe("of type standard view", () => {
    it("should render snapshot", () => {
      const { asFragment } = render(
        <ApplicationCard
          thumbnail={thumbnail}
          description={description}
          updatedAt={updatedAt}
          status={status}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("should open a menu when clicked", async () => {
      const { getByTestId, findByText } = render(
        <ApplicationCard
          thumbnail={thumbnail}
          description={description}
          updatedAt={updatedAt}
          status={status}
        />
      );
      fireEvent.click(getByTestId("moreButtonStandard"));
      expect(await findByText(/delete/i)).toBeInTheDocument();
      expect(await findByText(/archive/i)).toBeInTheDocument();
    });
  });
});
