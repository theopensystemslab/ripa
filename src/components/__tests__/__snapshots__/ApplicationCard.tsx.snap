// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application Card Component of type list view should render snapshot 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-container"
  >
    <div
      class="MuiListItem-root makeStyles-listedApplication-10 MuiListItem-gutters MuiListItem-secondaryAction"
      data-testid="listViewCard"
    >
      <div
        class="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          description
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          Last edited over 50 years ago
        </p>
      </div>
    </div>
    <div
      class="MuiListItemSecondaryAction-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        data-testid="moreButtonList"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`Application Card Component of type standard view should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-application-32 MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="standardCard"
  >
    <div
      class="MuiCardMedia-root makeStyles-thumbnail-37"
      style="background-image: url(pic.png);"
    />
    <div
      class="MuiCardContent-root makeStyles-content-36"
    >
      <div
        class="MuiBox-root MuiBox-root-46"
      >
        <strong>
          description
        </strong>
      </div>
      <div
        class="MuiBox-root MuiBox-root-47"
      >
        Last edited over 50 years ago
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <div
        class="MuiBox-root MuiBox-root-48"
      >
        Submitted
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        data-testid="moreButtonStandard"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
